### PR TITLE
[python] Check for downgraded categorical write

### PR DESCRIPTION
**Pull request by johnkerl**
**Issue and/or context:** #1739

**Changes:**

**Notes for Reviewer:**

### CODE REVIEW

File: apis/python/tests/test_dataframe.py [hunk 1]
```diff
@@ -139,11 +139,17 @@ def test_dataframe_with_enumeration(tmp_path):
...
+            data["foo"] = ["a", "bb", "ccc", "bb", "a"]
+            data["bar"] = ["cat", "dog", "cat", "cat", "cat"]
+            with pytest.raises(ValueError):
+                sdf.write(pa.Table.from_pydict(data))
+        else:
+            data["foo"] = pd.Categorical(["a", "bb", "ccc", "bb", "a"])
+            data["bar"] = pd.Categorical(["cat", "dog", "cat", "cat", "cat"])
+            sdf.write(pa.Table.from_pydict(data))
+            assert sdf.enumeration("foo") == enums["enmr1"]
+            assert sdf.enumeration("bar") == enums["enmr2"]
```
Comment (nguyenv) [inline apis/python/tests/test_dataframe.py:152]: I don't think this needs to be parameterized.

```suggestion
          data["foo"] = ["a", "bb", "ccc", "bb", "a"]
          data["bar"] = ["cat", "dog", "cat", "cat", "cat"]
          with pytest.raises(ValueError):
              sdf.write(pa.Table.from_pydict(data))

          data["foo"] = pd.Categorical(["a", "bb", "ccc", "bb", "a"])
          data["bar"] = pd.Categorical(["cat", "dog", "cat", "cat", "cat"])
          sdf.write(pa.Table.from_pydict(data))
          assert sdf.enumeration("foo") == enums["enmr1"]
          assert sdf.enumeration("bar") == enums["enmr2"]
```

**Comment by johnkerl** [comment]
@nguyenv ready for round two!

File: apis/python/src/tiledbsoma/_dataframe.py [hunk 2]
```diff
@@ -388,6 +388,11 @@ def write(
                 containing all columns, including the index columns. The schema for the values must
                 match the schema for the :class:`DataFrame`.
 
+                If a column is of categorical type in the schema and a flattened/non-categorical
+                column is presented for data on write, a ``ValueError`` is raised.  If a column is
+                of non-categorical type in the schema and a categorical column is presented for data
+                on write, the data are auto-flattened on behalf of the user.
```
Comment (bkmartinjr) [inline apis/python/src/tiledbsoma/_dataframe.py:394]: I don't think the term `auto-flattened` is a good way to express the operation (the term is used in Pandas to imply index flattening, etc).

Suggestion:

"..., the data are written as an array of category values, and the category type information is not saved."

Or something like that (my suggestion is also not quite right - hoping we can iterate)

File: apis/python/tests/test_dataframe.py [hunk 3]
```diff
@@ -122,15 +122,15 @@ def test_dataframe_with_float_dim(tmp_path, arrow_schema):
     assert sdf.index_column_names == ("bar",)
 
 
-def test_dataframe_with_enumeration(tmp_path):
+@pytest.mark.parametrize("downgrade", [False, True])
```
Comment (bkmartinjr) [inline apis/python/tests/test_dataframe.py:125]: what is the `downgrade` param?

**Review by bkmartinjr — CHANGES_REQUESTED** [review]
Mostly nits.  
* Not sure why there are R CI changes included in this PR
* Docstring suggestion (imperfect suggestion)
* Nit on unit test
Small stuff!

File: apis/python/tests/test_dataframe.py (hunk 3 above)
Comment (johnkerl) [reply apis/python/tests/test_dataframe.py:125]: oops went away on previous code-review-feedback param -- will remove

File: apis/python/src/tiledbsoma/_dataframe.py (hunk 2 above)
Comment (johnkerl) [reply apis/python/src/tiledbsoma/_dataframe.py:394]: > hoping we can iterate

Of course!!

**Review by nguyenv — APPROVED** [review]

**Review by bkmartinjr — APPROVED** [review]

**Pull request merged**